Add initial project structure with CMake configuration, Clang format …

Katze719 opened the pull request:
…and tidy settings, GitHub workflows, and integration tests for Windows serial communication bindings

Katze719 committed:
Add initial project structure with CMake configuration, Clang format and tidy settings, GitHub workflows, and integration tests for Windows serial communication bindings

Katze719 committed:
Add compile definitions for Windows bindings in CMake configuration

Katze719 committed:
Disable Google Mock in CMake configuration for Windows bindings

Katze719 committed:
Remove embed_binary.sh script and refactor applyLineSettings function to return structured error messages in serial_open.cpp and serial_read.cpp.

Katze719 committed:
Add readExact function to improve serial reading in tests; update expectations in SerialArduinoTest

Katze719 committed:
Add virtual COM port setup in GitHub workflows and update SerialArduinoTest to use COM5. This includes installing the com0com driver, creating a virtual COM pair, and starting an echo loop for testing. Adjusted the default serial port in tests to COM5 and improved error handling for port opening.

Katze719 committed:
Refactor virtual COM port setup in GitHub workflows to enhance error handling and improve script execution. Updated paths for locating setupc.exe and com0com.inf, and added a sleep delay after starting the echo loop for stability.

Katze719 committed:
Refactor virtual COM port setup in Deno and C++ GitHub workflows to improve path handling and error reporting. Updated script to search for setupc.exe and com0com.inf exclusively within the Chocolatey package directory, enhancing efficiency and reliability.

Katze719 committed:
Enhance virtual COM port setup in GitHub workflows by implementing a more efficient and deterministic search for setupc.exe and com0com.inf. Updated error handling to check multiple candidate paths and limit search depth, improving reliability in locating necessary files.

Katze719 committed:
Update CMake configuration to use cmake-git-versioning v1.1.0, add CMake presets for Windows builds, and enhance test source organization. Introduce new tests for serial operations and error handling, including serial_close, serial_open, serial_read, and serial_write. Add error capture utility for improved error reporting in tests.

Katze719 committed:
Enhance GitHub workflows by introducing a new build binary workflow for Windows, replacing the previous Deno DLL build and C++ test workflows. Add CMake presets for Visual Studio 2022 and update the JSR package structure to include a new binary export format. Refactor the JSR metadata to improve organization and include additional source files. Remove outdated workflows to streamline the CI process.

Katze719 committed:
Refactor JSR package structure by removing outdated metadata files and introducing a new versioning script. Update GitHub workflows to prepare files for JSR publishing, including changes to directory structure and file paths. Add necessary files such as LICENSE, README.md, and jsr.json for improved package organization and usability.

Katze719 committed:
Update GitHub workflow to streamline test build process by removing the target specification for C++ test binaries, simplifying the build command.

Katze719 committed:
Format jsr.json for improved readability and organization, enhancing the structure of the package metadata.

Katze719 committed:
Update CMake configuration to disable Google Mock support and remove related includes from test files, streamlining the testing setup.

Katze719 committed:
Update handle validation in serial operations to include checks against std::numeric_limits<int> for improved error handling and robustness.

Katze719 committed:
Remove the GNU General Public License text from the LICENSE file, streamlining the file for better clarity and focus on essential information.

Katze719 committed:
Update license field in jsr.json to specify GPL-3.0-only, ensuring compliance with licensing requirements.

Mqxx committed:
Update README.md

Mqxx committed:
Update jsr.json

Mqxx committed:
Update LICENSE

Mqxx reviewed (commented):
Will check if the windows side works.

Katze719 committed:
Refactor serial port handling to support UTF-8 input by adding utf8ToWide conversion function. Update tests to reflect changes in port type from wchar_t to char, ensuring compatibility with new implementation.

Mqxx commented on .github/workflows/build_binary.yml:
```diff
@@ -0,0 +1,101 @@
+          $dll = Get-ChildItem -Recurse -Path build -Filter cpp_bindings_windows.dll -File | Select-Object -First 1
+          if (-not $dll) { throw "cpp_bindings_windows.dll not found under build/" }
+          New-Item -ItemType Directory -Force -Path build/out | Out-Null
+          Copy-Item -Force $dll.FullName -Destination build/out/libcpp_bindings_windows.dll
```
Isn't the `lib` prefix specific to Linux? As far as I know only Linux does prefix the libs with `lib`.

Katze719 replied:
disabled the lib prefix for windows

Mqxx commented on .github/workflows/publish_jsr.yml:
```diff
@@ -0,0 +1,69 @@
+          if (-not (Test-Path $dll)) { throw "DLL not found: $dll" }
+          New-Item -ItemType Directory -Force -Path jsr/bin | Out-Null
+          ./jsr/scripts/embed_binary.ps1 $dll jsr/bin windows-x86_64
+          Copy-Item -Force LICENSE jsr/LICENSE
```
```suggestion
```

The package should have its own LICENSE file, even if its the same file as in the root.

You can use this LICENSE file: https://github.com/Serial-IO/cpp-bindings-linux/blob/main/jsr/LICENSE

Mqxx commented on jsr/scripts/embed_binary.ps1:
```diff
@@ -0,0 +1,40 @@
+$filename = [System.IO.Path]::GetFileName($BinaryPath)
+
+# Copy dll into the package bin folder with a stable name
+Copy-Item -Force -LiteralPath $BinaryPath -Destination (Join-Path $JsrBinPath "x86_64.dll")
```
The name should be the filename from the path. Not hardcoded to `x86_64.dll`.

Also copying and embedding the binary should be two separate scripts like here: https://github.com/Serial-IO/cpp-bindings-linux/tree/main/jsr/scripts

Mqxx committed:
fix: refactor

Mqxx committed:
Update publish_jsr.yml

Mqxx committed:
Update publish_jsr.yml

Mqxx committed:
Update publish_jsr.yml

Mqxx committed:
Update build_binary.yml

Mqxx committed:
Update build_binary.yml

Mqxx committed:
Update build_binary.yml

Katze719 commented on .github/workflows/build_binary.yml:
```diff
@@ -0,0 +1,111 @@
+          path: build/out/libcpp_bindings_windows.dll
+
+      - name: 'Create GitHub Release'
+        if: startsWith(github.ref, 'refs/tags/') && steps.check-tag.outputs.IS_VALID_PACKAGE_VERSION == 'true'
```
```suggestion
        if: github.ref_type == 'tag' && steps.check-tag.outputs.IS_VALID_PACKAGE_VERSION == 'true'
```

Mqxx committed:
Update .github/workflows/build_binary.yml

Co-authored-by: Katze719 <38188106+Katze719@users.noreply.github.com>

Katze719 committed:
Refactor artifact naming and paths in CI workflows

- Updated artifact names from 'libcpp_bindings_windows' to 'cpp_bindings_windows' for consistency across workflows.
- Adjusted file paths in build and publish scripts to reflect the new artifact naming convention.

Mqxx commented on src/detail/win32_helpers.hpp:
Could you refactor this big file into multiple files and place them into a folder at here called `win32`?

Katze719 replied:
will be refactored with `cpp-core v1.1.0`

Mqxx commented on src/serial_open.cpp:
File is really long. Could you please refactor this file and extract all the functions into their corresponding file.

Katze719 replied:
will be refactored with `cpp-core v1.1.0`

Mqxx commented on src/serial_read.cpp:
This file is also extremely long.

Katze719 replied:
will be refactored with `cpp-core v1.1.0`

Katze719 commented:
waiting for `cpp-core v1.1.0` release

Mqxx committed:
Create block_pr_check_labels.yml

Mqxx committed:
Update block_pr_check_labels.yml

Mqxx commented:
> waiting for `cpp-core v1.1.0` release

Version `1.1.0` was merged 👍

Katze719 committed:
Update CMakeLists.txt and refactor serial communication code

- Changed cpp_core package version from 'main' to 'v1.1.0' in CMakeLists.txt.
- Refactored serial_close, serial_open, serial_read, and serial_write implementations to use new validation and error handling mechanisms.
- Removed deprecated status code checks and replaced them with cpp_core's validation functions for better error management.
- Improved readability and maintainability of the serial communication functions by utilizing modern C++ practices.

Katze719 committed:
Refactor handle creation in serial_open to improve error handling

- Introduced a raw_handle variable to capture the result of CreateFileW.
- Updated UniqueHandle initialization to handle INVALID_HANDLE_VALUE by normalizing it to nullptr, enhancing error management in serial communication.

Mqxx commented on src/serial_open.cpp:
```diff
@@ -0,0 +1,190 @@
+
+namespace
+{
+auto utf8ToWide(const char *utf8) -> std::wstring
+{
+    if (utf8 == nullptr || utf8[0] == '\0')
+    {
+        return {};
+    }
+    const int needed = MultiByteToWideChar(CP_UTF8, 0, utf8, -1, nullptr, 0);
+    if (needed <= 0)
+    {
+        return {};
+    }
+    std::wstring out(static_cast<size_t>(needed), L'\0');
+    const int written = MultiByteToWideChar(CP_UTF8, 0, utf8, -1, out.data(), needed);
+    if (written <= 0)
+    {
+        return {};
+    }
+    if (!out.empty() && out.back() == L'\0')
+    {
+        out.pop_back();
+    }
+    return out;
+}
```
Please refactor to own file

Katze719 replied:
is only needed in serial_open.cpp, and this function remains in the serial_open.cpp to enforce local linkage, so the function does not get exposed to other files (private function)

Mqxx replied:
It does not matter if it's "only needed" in the `serialOpen` function, you *could* still use it in other functions/methods in the future, 'cause its generic enough to be used in other functions/methods. For the "so the function does not get exposed to other files (private function)", you can just exclude the method from the umbrella header file.

Katze719 replied:
no, if the function gets included in a file it will be available in every file that includes that one, its not a module system

if i include A into B, and i include B into C, A will be available in C, and i do not want that

Mqxx replied:
Okay, I rephrase my question: Why does this function has to be private? Why can't you just treat it as a normal function that everyone can use? I don't see a particular reason for it to be private...

Katze719 replied:
i dont feel like hardening the function to support every use case, instead i focus on keeping the function simple and perfect for my current use case, if i expose the function to the user, i can not rename the function anymore without having a breaking change.

and it would force users to sift through functions they should not call

Mqxx replied:
> i dont feel like hardening the function to support every use case

Why does the function need to support every use case? What use cases anyways?

> instead i focus on keeping the function simple and perfect for my current use case

The function can still be simple and the same, no changes needed.

> if i expose the function to the user, i can not rename the function anymore without having a breaking change.

Why would you need to rename it? I don't see a particular reason, for it to be renamed. This is more likely speculation, and if there is a reason for it to be renamed, you can just do it and mark it as deprecated and replace it in the future.

For me it currently looks like "ohh I don't want to make it public, because then I have to properly maintain it...". And if this is not the case, there is no reason for the function to be private...

Mqxx commented on src/serial_open.cpp:
```diff
@@ -0,0 +1,190 @@
+    return out;
+}
+
+auto normalizePortPath(const wchar_t *port) -> std::wstring
+{
+    std::wstring p(port);
+    if (p.rfind(L"\\\\.\\", 0) == 0)
+    {
+        return p;
+    }
+    if (p.rfind(L"COM", 0) == 0 || p.rfind(L"com", 0) == 0)
+    {
+        return L"\\\\.\\" + p;
+    }
+    return p;
+}
```
Please refactor to own file

Katze719 replied:
is only needed in serial_open.cpp, and this function remains in the serial_open.cpp to enforce local linkage, so the function does not get exposed to other files (private function)

Mqxx replied:
Same as above comment.

Mqxx commented on src/serial_read.cpp:
```diff
@@ -0,0 +1,221 @@
+    return 1;
+}
+
+auto readSome(HANDLE handle, unsigned char *dst, int size, int timeout_ms) -> int
```
Is this a utility function only for `serialRead`?

Mqxx commented on src/serial_write.cpp:
```diff
@@ -0,0 +1,95 @@
+
+namespace
+{
+auto writeSome(HANDLE handle, const void *src, int size, int timeout_ms) -> int
```
Same here, is this just a utility function for `serialWrite`?
